### PR TITLE
Simple script for dataset conversion

### PR DESCRIPTION
**Pull request by vpchung**
Adding script for converting current CSBC/PS-ON "datasets" (Synapse Folders) into the Dataset entity type.

Similar steps from the script can be implemented into the curation workflow once CCKP is configured to support Datasets.  For now, the curation workflow will continue to upload the dataset manifests as Synapse Folders.

### CODE REVIEW

**Comment by thomasyu888** [comment]
I can see this as a synapseutils function!

File: python/convert_to_dataset.py [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+def login():
+    """Log into Synapse. If env variables not found, prompt user.
+
+    Returns:
+        syn: Synapse object
+    """
+    try:
+        syn = synapseclient.login(
+            os.getenv('SYN_USERNAME'),
+            apiKey=os.getenv('SYN_APIKEY'),
```
Comment (thomasyu888) [inline python/convert_to_dataset.py]: Don't use API key - use synapse PAT

File: python/convert_to_dataset.py [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+        list of `DatasetItem`s
+    """
+    files = syn.getChildren(folder_id)
+    return [
+        {'entityId': f.get('id'), 'versionNumber': f.get('versionLabel')}
+        for f in files
+    ]
+
+
+def convert_folder_to_dataset(syn, dataset_name, schema, dataset_items):
```
Comment (thomasyu888) [inline python/convert_to_dataset.py]: The name of this function is semi misleading.  I thought you were literally taking a Folder entity and converting the Folder Entity to a Dataset.  But you're creating a Dataset with the same name as the folder name.  `_create_dataset()` is more clear.

File: python/convert_to_dataset.py [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+        folder_id = folder.get('id')
+        files = convert_data_files(syn, folder_id)
+        try:
+            convert_folder_to_dataset(syn,
+                                      dataset_name=folder.get('name'),
+                                      schema=col_ids,
+                                      dataset_items=files)
+            folder_count += 1
+            file_count += len(files)
+        except synapseclient.core.exceptions.SynapseHTTPError:
```
Comment (thomasyu888) [inline python/convert_to_dataset.py]: Just out of curiosity - what are same of the errors you have gotten?

File: python/convert_to_dataset.py (hunk 1 above)
Comment (vpchung) [reply python/convert_to_dataset.py]: I know, it's already been noted in my TODO on line 6.

File: python/convert_to_dataset.py (hunk 2 above)
Comment (vpchung) [reply python/convert_to_dataset.py]: Sounds good, thanks!

File: python/convert_to_dataset.py (hunk 3 above)
Comment (vpchung) [reply python/convert_to_dataset.py]: The function is assuming that all immediate children of the Folder entity are File entities.  So the error I came across was when the Dataset was being created and not all of the DatasetItems are File entities.

That said, I think I need to move my assumption from line 66.  Thanks for bringing that to my attention :sweat_smile:

File: python/convert_to_dataset.py (hunk 1 above)
Comment (vpchung) [reply python/convert_to_dataset.py]: On this note, the [docs](https://python-docs.synapse.org/build/html/index.html#synapseclient.Synapse.login) should be updated to reflect this change, e.g. I didn't realize PAT is already accepted.

Comment (thomasyu888) [reply python/convert_to_dataset.py]: `authToken` is the PAT - I think the docs need to be improved.

Comment (vpchung) [reply python/convert_to_dataset.py]: Oh, d'oh. In retrospect, that makes sense.

Thanks!  I'll implement the update.

File: python/convert_to_dataset.py [hunk 4]
```diff
@@ -96,6 +96,17 @@ def _create_dataset(syn, dataset_name, schema, dataset_items):
     syn.restPOST("/entity", json.dumps(req_body))
 
 
+def _reset_datasets(syn):
+    """Remove all current datasets."""
+    curr_datasets = [i.get('id') for i in syn.restPOST(
+        "/entity/children", json.dumps({'parentId': PARENT_ID, "includeTypes": ["dataset"]})).get('page')]
```
Comment (thomasyu888) [inline python/convert_to_dataset.py:95]: you can do `syn.getChildren(PARENT_ID, includeTypes=["dataset"])`

File: python/.envTemplate [hunk 5]
```diff
@@ -3,10 +3,11 @@
...
 ## Used for logging in to Synapse
-# For security purposes, you should provide your apikey rather than
-# your password. Your apikey can be found on My Dashboard > Settings.
-SYN_USERNAME=username
-SYN_APIKEY="apikey"
+# For security purposes, you should provide your personal access token
+# (PAT) rather than your password. You can generate PATs on Synapse by
+# going to Account Settings > Manage Personal Access Tokens.
+export SYN_USERNAME=username
+export SYN_PAT="PAT"
```
Comment (thomasyu888) [inline python/.envTemplate]: A cool feature is that the synapsePythonClient actually looks for the env variable `SYNAPSE_AUTH_TOKEN`.  So by doing:

```
export SYNAPSE_AUTH_TOKEN="PAT"
```

the client will auto login by just doing `synapseclient.login()`. You technically don't need the username, but perhaps it helps you remember whose PAT it is.

File: python/convert_to_dataset.py [hunk 6]
```diff
@@ -126,8 +126,13 @@ def main():
...
             folder_count += 1
             file_count += len(files)
-        except synapseclient.core.exceptions.SynapseHTTPError:
-            needs_review.append(folder_id)
+        except synapseclient.core.exceptions.SynapseHTTPError as err:
+
+            # 409 Client Error is from adding a dataset with an existing name.
+            # REVIEW. Double-check if this is how Synapse errors should be
+            #         caught.
+            if not str(err).startswith("409 Client Error:"):
```
Comment (thomasyu888) [inline python/convert_to_dataset.py]:
```suggestion
            if not err.response.status_code == 409:
```

I think that should work

**Comment by vpchung** [comment]
Closing, as we may/may not use Datasets in the future.

**Pull request closed without merging**